binder/src/main/scala/PanamaCIRCTConverter.scala

```diff
@@ -721,6 +725,114 @@ class PanamaCIRCTConverter extends CIRCTConverter {
         .withOperand( /* src */ src.value)
         .build()
     }
+
+    case class RecursiveTypeProperties(isPassive: Boolean, containsAnalog: Boolean)
+
+    def recursiveTypeProperties(tpe: fir.Type): RecursiveTypeProperties = {
+      tpe match {
+        case fir.ClockType | fir.ResetType | fir.AsyncResetType | _: fir.SIntType | _: fir.UIntType =>
+          RecursiveTypeProperties(true, false)
+        case _:      fir.AnalogType => RecursiveTypeProperties(true, true)
+        case bundle: fir.BundleType =>
+          bundle.fields.foldLeft(RecursiveTypeProperties(true, false)) {
+            case (properties, field) =>
+              val fieldProperties = recursiveTypeProperties(field.tpe)
+              RecursiveTypeProperties(
+                properties.isPassive && fieldProperties.isPassive && field.flip == fir.Flip,
+                properties.containsAnalog || fieldProperties.containsAnalog
+              )
+          }
+        case _:      fir.PropertyType => RecursiveTypeProperties(true, false)
+        case vector: fir.VectorType   => recursiveTypeProperties(vector.tpe)
+      }
+    }
+
+    sealed trait Flow
+    object Flow {
+      final case object None extends Flow
+      final case object Source extends Flow
+      final case object Sink extends Flow
+      final case object Duplex extends Flow
+    }
+
+    def swapFlow(flow: Flow): Flow = flow match {
+      case Flow.None   => Flow.None
+      case Flow.Source => Flow.Sink
+      case Flow.Sink   => Flow.Source
+      case Flow.Duplex => Flow.Duplex
+    }
+
+    def foldFlow(value: Reference.Value, acc: Flow = Flow.Source): Flow = {
+      circt.firrtlValueFoldFlow(
+        value.value,
+        acc match {
+          case Flow.None   => 0
+          case Flow.Source => 1
+          case Flow.Sink   => 2
+          case Flow.Duplex => 3
+        }
+      ) match {
+        case 0 => Flow.None
+        case 1 => Flow.Source
+        case 2 => Flow.Sink
+        case 3 => Flow.Duplex
+      }
+    }
+
+    def emitInvalidate(value: Reference.Value, loc: MlirLocation): Unit = emitInvalidate(value, loc, foldFlow(value))
```

Comment:
I'm getting scared about binder require `flow`, which means: Rather than plain emit IR, Chisel need to query the states from MFC. I personally prefer we should improve the semantic of FIRRTL.